rclcpp/include/rclcpp/executors/executor_entities_collection.hpp

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__EXECUTORS__EXECUTOR_ENTITIES_COLLECTION_HPP_
+#define RCLCPP__EXECUTORS__EXECUTOR_ENTITIES_COLLECTION_HPP_
+
+#include <deque>
+#include <unordered_map>
+#include <vector>
+
+#include <rclcpp/any_executable.hpp>
+#include <rclcpp/node_interfaces/node_base.hpp>
+#include <rclcpp/callback_group.hpp>
+#include <rclcpp/executors/executor_notify_waitable.hpp>
+#include <rclcpp/visibility_control.hpp>
+#include <rclcpp/wait_result.hpp>
+#include <rclcpp/wait_set.hpp>
+
+namespace rclcpp
+{
+namespace executors
+{
+
+template<typename EntityValueType>
+struct CollectionEntry
+{
+  typename EntityValueType::WeakPtr entity;
+  rclcpp::CallbackGroup::WeakPtr callback_group;
+};
+
+template<typename CollectionType>
+void update_entities(
+  const CollectionType & update_from,
+  CollectionType update_to,
+  std::function<void(typename CollectionType::mapped_type::EntitySharedPtr)> on_added,
+  std::function<void(typename CollectionType::mapped_type::EntitySharedPtr)> on_removed
```

Comment:
there's no `mapped_type` definition, this should just be
```suggestion
  std::function<void(typename CollectionType::EntitySharedPtr)> on_removed
```